balancer/pickfirst/pickfirstleaf/pickfirstleaf_ext_test.go

```diff
@@ -851,6 +850,245 @@ func (s) TestPickFirstLeaf_EmptyAddressList(t *testing.T) {
+	opts := []grpc.DialOption{
+		grpc.WithDefaultServiceConfig(fmt.Sprintf(`{"loadBalancingConfig": [{"%s":{}}]}`, pickfirstleaf.Name)),
+		grpc.WithContextDialer(dialer.DialContext),
+	}
+	cc, rb, bm := setupPickFirstLeaf(t, 2, opts...)
+	addrs := bm.resolverAddrs()
+	holds := bm.holds(dialer)
+	rb.UpdateState(resolver.State{Addresses: addrs})
+	cc.Connect()
+
+	testutils.AwaitState(ctx, t, cc, connectivity.Connecting)
+
+	// Verify that the first server is contacted and the second is not.
+	if holds[0].Wait(ctx) != true {
+		t.Fatalf("Timeout waiting for server %d with address %q to be contacted", 0, addrs[0])
+	}
+	if holds[1].IsStarted() != false {
+		t.Fatalf("Server %d with address %q contacted unexpectedly", 1, addrs[1])
+	}
+
+	timerCh <- struct{}{}
+
+	// Second connection attempt is successful.
+	if holds[1].Wait(ctx) != true {
+		t.Fatalf("Timeout waiting for server %d with address %q to be contacted", 1, addrs[1])
+	}
+	holds[1].Resume()
+	testutils.AwaitState(ctx, t, cc, connectivity.Ready)
+}
+
+// TestPickFirstLeaf_HappyEyeballs_TFThenTimerFires tests the pickfirst balancer
+// by causing a SubConn to fail and then jumping to the 3rd SubConn after the
+// happy eyeballs timer expires.
+func (s) TestPickFirstLeaf_HappyEyeballs_TFThenTimerFires(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	timerMu := sync.Mutex{}
+	timerCh := make(chan struct{})
+	originalTimer := pfinternal.TimeAfterFunc
+	pfinternal.TimeAfterFunc = func(_ time.Duration, f func()) *time.Timer {
+		// Set a really long expiration to prevent it from triggering
+		// automatically.
+		timer := time.AfterFunc(time.Hour, f)
+		timerMu.Lock()
+		ch := timerCh
+		timerMu.Unlock()
+		go func() {
+			select {
+			case <-ctx.Done():
+			case <-ch:
+			}
+			timer.Reset(0)
+		}()
+		return timer
+	}
+
+	defer func() {
+		pfinternal.TimeAfterFunc = originalTimer
+	}()
+
+	dialer := testutils.NewBlockingDialer()
+	opts := []grpc.DialOption{
+		grpc.WithDefaultServiceConfig(fmt.Sprintf(`{"loadBalancingConfig": [{"%s":{}}]}`, pickfirstleaf.Name)),
+		grpc.WithContextDialer(dialer.DialContext),
+	}
+	cc, rb, bm := setupPickFirstLeaf(t, 3, opts...)
+	addrs := bm.resolverAddrs()
+	holds := bm.holds(dialer)
+	rb.UpdateState(resolver.State{Addresses: addrs})
+	cc.Connect()
+
+	testutils.AwaitState(ctx, t, cc, connectivity.Connecting)
+
+	// Verify that only the first server is contacted.
+	if holds[0].Wait(ctx) != true {
+		t.Fatalf("Timeout waiting for server %d with address %q to be contacted", 0, addrs[0])
+	}
+
+	// Ensure no other servers are contacted.
+	if holds[1].IsStarted() != false {
+		t.Fatalf("Server %d with address %q contacted unexpectedly", 1, addrs[1])
+	}
+	if holds[2].IsStarted() != false {
+		t.Fatalf("Server %d with address %q contacted unexpectedly", 2, addrs[2])
+	}
+
+	// Replace the timer channel so that the old timers don't attempt to read
+	// messages pushed next.
```

Comment:
Old timers should get canceled when subsequent subchannels are created, right? Why do we need to do this?

Reply:
I see what you are saying. That seems better to me, unless it is too much work.

Reply:
Refactored to have the internal.TimeAfterFunc return a `cancelFunc()` instead of a timer. This allowed the test to stop the timer when `pickfirst` cancels the timer. I also created a helper function to return a timer function and a function to trigger the timer manually instead of having the tests write on channel.